File: docs/Claude-Control-Center-main/backend/app/routes/git_tree.py
from __future__ import annotations

"""Git Tree API — worktree management and commit graph with attribution."""

from flask import Blueprint, Response, abort, jsonify, request

import orjson

from app.services import github_repos_service, git_local_service, git_worktree_service

bp = Blueprint("git_tree", __name__, url_prefix="/api/git-tree")


def _orjson_response(data, status: int = 200) -> Response:
    return Response(
        orjson.dumps(data),
        status=status,
        mimetype="application/json",
    )


def _get_repo(repo_id: str):
    """Return (repo_dict, validated_path) or abort 404."""
    repos = github_repos_service.discover_repos()
    repo = next((r for r in repos if r.get("id") == repo_id), None)
    if repo is None:
        abort(404, description="Repo not found")
    roots = github_repos_service.list_roots()
    try:
        path = git_local_service.validate_repo_path(repo["path"], roots)
    except ValueError as exc:
        abort(400, description=str(exc))
    return repo, path


# ---------------------------------------------------------------------------
# Repo list (reuses github roots discovery)
# ---------------------------------------------------------------------------

@bp.get("/repos")
def list_repos():
    repos = github_repos_service.discover_repos()
    return _orjson_response(repos)


# ---------------------------------------------------------------------------
# Worktree endpoints
# ---------------------------------------------------------------------------

@bp.get("/all-worktrees")
def all_worktrees():
    """Return worktrees for all enabled repos."""
    repos = github_repos_service.discover_repos()
    roots = github_repos_service.list_roots()
    result = []
    for repo in repos:
        try:
            path = git_local_service.validate_repo_path(repo["path"], roots)
            wts = git_worktree_service.list_worktrees(path)
            for wt in wts:
                result.append({**wt, "repo_id": repo["id"], "repo_name": repo["name"]})
        except Exception:
            continue
    return _orjson_response(result)


@bp.get("/repos/<repo_id>/worktrees")
def list_worktrees(repo_id: str):
    _, path = _get_repo(repo_id)
    return _orjson_response(git_worktree_service.list_worktrees(path))


@bp.post("/repos/<repo_id>/worktrees")
def create_worktree(repo_id: str):
    _, path = _get_repo(repo_id)
    body = request.get_json(silent=True) or {}
    dest = (body.get("dest") or "").strip()
    branch = (body.get("branch") or "").strip()
    new_branch = bool(body.get("new_branch", False))
    if not dest or not branch:
        return jsonify({"error": "dest and branch are required"}), 400
    try:
        result = git_worktree_service.add_worktree(path, dest, branch, new_branch)
    except RuntimeError as exc:
        return jsonify({"error": str(exc)}), 422
    return _orjson_response(result, 201)


@bp.delete("/repos/<repo_id>/worktrees")
def delete_worktree(repo_id: str):
    _, path = _get_repo(repo_id)
    body = request.get_json(silent=True) or {}
    wt_path = (body.get("path") or "").strip()
    force = bool(body.get("force", False))
    if not wt_path:
        return jsonify({"error": "path is required"}), 400
    try:
        git_worktree_service.remove_worktree(path, wt_path, force=force)
    except RuntimeError as exc:
        return jsonify({"error": str(exc)}), 422
    return _orjson_response({"removed": True})


@bp.post("/repos/<repo_id>/worktrees/lock")
def lock_worktree(repo_id: str):
    _, path = _get_repo(repo_id)
    body = request.get_json(silent=True) or {}
    wt_path = (body.get("path") or "").strip()
    reason = (body.get("reason") or "").strip()
    if not wt_path:
        return jsonify({"error": "path is required"}), 400
    try:
        git_worktree_service.lock_worktree(path, wt_path, reason)
    except RuntimeError as exc:
        return jsonify({"error": str(exc)}), 422
    return _orjson_response({"locked": True})


@bp.post("/repos/<repo_id>/worktrees/unlock")
def unlock_worktree(repo_id: str):
    _, path = _get_repo(repo_id)
    body = request.get_json(silent=True) or {}
    wt_path = (body.get("path") or "").strip()
    if not wt_path:
        return jsonify({"error": "path is required"}), 400
    try:
        git_worktree_service.unlock_worktree(path, wt_path)
    except RuntimeError as exc:
        return jsonify({"error": str(exc)}), 422
    return _orjson_response({"unlocked": True})


# ---------------------------------------------------------------------------
# Commit graph
# ---------------------------------------------------------------------------

@bp.get("/repos/<repo_id>/graph")
def commit_graph(repo_id: str):
    _, path = _get_repo(repo_id)
    try:
        limit = int(request.args.get("limit", 100))
        limit = max(1, min(limit, 500))
    except ValueError:
        limit = 100
    commits = git_worktree_service.get_commit_graph(path, limit=limit)
    return _orjson_response(commits)

File: docs/Claude-Control-Center-main/frontend/src/api/gitTree.ts
import { apiFetch } from './client';
import type { GitCommit, GitRepo, GitWorktree, GitWorktreeWithRepo } from '../types';

export type { GitCommit, GitRepo, GitWorktree, GitWorktreeWithRepo };

export async function fetchGitRepos(): Promise<GitRepo[]> {
  return apiFetch('/api/git-tree/repos');
}

export async function fetchCommitGraph(repoId: string, limit = 100): Promise<GitCommit[]> {
  const raw = await apiFetch<Record<string, unknown>[]>(
    `/api/git-tree/repos/${encodeURIComponent(repoId)}/graph?limit=${limit}`
  );
  return raw.map((c) => ({
    hash: String(c.hash ?? ''),
    shortHash: String(c.short_hash ?? ''),
    subject: String(c.subject ?? ''),
    authorName: String(c.author_name ?? ''),
    authorEmail: String(c.author_email ?? ''),
    committedAt: String(c.committed_at ?? ''),
    refs: (c.refs as string[]) ?? [],
    parents: (c.parents as string[]) ?? [],
    body: String(c.body ?? ''),
    attribution: (c.attribution as GitCommit['attribution']) ?? 'unknown',
    attributionSources: (c.attribution_sources as string[]) ?? [],
    coAuthors: (c.co_authors as string[]) ?? [],
  }));
}

export async function fetchWorktrees(repoId: string): Promise<GitWorktree[]> {
  const raw = await apiFetch<Record<string, unknown>[]>(
    `/api/git-tree/repos/${encodeURIComponent(repoId)}/worktrees`
  );
  return raw.map(_mapWorktree);
}

export async function fetchAllWorktrees(): Promise<GitWorktreeWithRepo[]> {
  const raw = await apiFetch<Record<string, unknown>[]>('/api/git-tree/all-worktrees');
  return raw.map((r) => ({
    ..._mapWorktree(r),
    repoId: String(r.repo_id ?? ''),
    repoName: String(r.repo_name ?? ''),
  }));
}

function _mapWorktree(r: Record<string, unknown>): GitWorktree {
  return {
    path: String(r.path ?? ''),
    branch: r.branch != null ? String(r.branch) : null,
    headHash: String(r.head_hash ?? ''),
    isMain: Boolean(r.is_main),
    isDetached: Boolean(r.is_detached),
    isLocked: Boolean(r.is_locked),
    lockReason: r.lock_reason != null ? String(r.lock_reason) : null,
    prunable: Boolean(r.prunable),
    isBare: Boolean(r.is_bare),
  };
}

export async function createWorktree(
  repoId: string,
  payload: { dest: string; branch: string; newBranch: boolean }
): Promise<{ path: string; branch: string }> {
  return apiFetch(`/api/git-tree/repos/${encodeURIComponent(repoId)}/worktrees`, {
    method: 'POST',
    body: JSON.stringify({
      dest: payload.dest,
      branch: payload.branch,
      new_branch: payload.newBranch,
    }),
  });
}

export async function removeWorktree(
  repoId: string,
  payload: { path: string; force: boolean }
): Promise<{ removed: boolean }> {
  return apiFetch(`/api/git-tree/repos/${encodeURIComponent(repoId)}/worktrees`, {
    method: 'DELETE',
    body: JSON.stringify(payload),
  });
}

export async function lockWorktree(
  repoId: string,
  payload: { path: string; reason?: string }
): Promise<{ locked: boolean }> {
  return apiFetch(`/api/git-tree/repos/${encodeURIComponent(repoId)}/worktrees/lock`, {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export async function unlockWorktree(
  repoId: string,
  payload: { path: string }
): Promise<{ unlocked: boolean }> {
  return apiFetch(`/api/git-tree/repos/${encodeURIComponent(repoId)}/worktrees/unlock`, {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}
